### ISSUE TITLE
Make Cache Invalidation Work per Function Instead of per File

### ISSUE THREAD

**Issue by DannyWeitekamp** [numba/numba#9316]
TLDR: A caching issue and potential solution.

Something that is quite annoying is that if you edit anything in a source file every function decorated with @njit(cache=True) will have its cache invalidated, causing all of those functions to recompile. In a large project this can significantly slow down development since you need to wait several seconds or even minutes after every edit. It would be nice if only the functions effected by an edit needed to recompile. I'm especially worried about how this issue becoming even more difficult in the near future after this PR (#8396) is implemented which will makes it so that cache invalidations cascades through dependencies. This is an important feature, although I selfishly worry that when this is incorporated nearly my entire project is going to need to unnecessarily recompile every time I tweak a file with any functions heavily elsewhere in the project, even if the implementation of those dependencies are left unchanged. 

I'm hoping to open up a discussion about how this issue could be addressed and offer a prototype solution. A big part of why this issue is hard is that functions definitions are not self-contained in their bodies, they can leverage constants and other functions which are in a function's `__global__` scope and `__closure__`. The current solution of timestamping files seems more straightforward than checking the whole `__global__` dict for each function. But why not instead just check the elements of `__global__`  which are referenced in the function, i.e. the elements which overlap with `func.__code__.co_names`. 

Below is a monkey patched solution along these lines. The function is stamped by a hash of its `func.__code__.co_code`, and the subset of  it's `__globals__` dict which is explicitly referenced in the function body.

```python
import cloudpickle
from numba.core.caching import Cache, _SourceFileBackedLocatorMixin, _CacheLocator, _UserProvidedCacheLocator, CompileResultCacheImpl
from numba.core.dispatcher import Dispatcher
from numba import njit
from numba import config
import os
import types as py_types
import functools
from copy import copy
import hashlib

import time
class PrintElapse():
    def __init__(self, name):
        self.name = name
    def __enter__(self):
        self.t0 = time.time_ns()/float(1e6)
    def __exit__(self,*args):
        self.t1 = time.time_ns()/float(1e6)
        print(f'{self.name}: {self.t1-self.t0:.6f} ms') 

class _PreciseCacheLocator(_UserProvidedCacheLocator):
    def __init__(self, py_func, py_file):
        self._py_func = py_func
        cache_subpath = self.get_suitable_cache_subpath(py_file)
        self._cache_path = os.path.join(config.CACHE_DIR, cache_subpath)

        with PrintElapse("hashing time"):
            code =  self._py_func.__code__
            glbs = self._py_func.__globals__

            # Get __globals__ referenced in the function body
            _vars = code.co_names
            used_globals = {k:glbs[k] for k in _vars if k in glbs}

            # Hash the function bytecode and subset of globals
            func_bytes = code.co_code + cloudpickle.dumps(used_globals)
            self._func_hash = hashlib.sha256(func_bytes).hexdigest()
    
    def get_source_stamp(self):
        return self._func_hash

    def get_disambiguator(self):
        return self._func_hash[:10]

    @classmethod
    def from_function(cls, py_func, py_file):
        # if not config.CACHE_DIR:
        #     return
        self = cls(py_func, py_file)
        return self
        

class PreciseCacheImpl(CompileResultCacheImpl):
    _locator_classes = [
        _PreciseCacheLocator,
        *CompileResultCacheImpl._locator_classes
    ] 

class PreciseCache(Cache):
    """
    Implements Cache that saves and loads CompileResult objects.
    """
    _impl_class = PreciseCacheImpl


# Monkey Patch Dispatcher
def enable_caching(self):
    self._cache = PreciseCache(self.py_func)
Dispatcher.enable_caching = enable_caching
```
We can then execute something like this
```python
GLOBAL_VAL1 = 1
GLOBAL_VAL2 = 2
GLOBAL_VAL3 = 3

@njit(cache=True)
def foo():
    return GLOBAL_VAL1

@njit(cache=True)
def bar():
    return GLOBAL_VAL2

def gen_baz():
    x = GLOBAL_VAL3+0    
    @njit(cache=True)
    def baz():
        return x
    return baz

baz = gen_baz()
print(foo(), bar(), baz())
print("FOO", "HIT" if len(foo._cache_misses) == 0 else "MISS")
print("BAR", "HIT" if len(bar._cache_misses) == 0 else "MISS")
print("BAZ", "HIT" if len(baz._cache_misses) == 0 else "MISS")
```
on the first run the cache misses as expected, and a quick timing of the hashing parts show that this method seems fast enough to be a viable solution.
```
hashing time: 0.034180 ms
hashing time: 0.012939 ms
hashing time: 0.006592 ms
1 2 3
FOO MISS
BAR MISS
BAZ MISS
```
Subsequent runs without any edits hit
```
...
1 2 3
FOO HIT
BAR HIT
BAZ HIT
```
And if I change any of the  `GLOBAL_VAL` values then only the functions that depends on those globals will recompile.  This seems to work (at least in this example) even if the jitted function was  produced in a closure.
```python
GLOBAL_VAL1 = 7
GLOBAL_VAL2 = 2
GLOBAL_VAL3 = 8
```
```
7 2 8
FOO MISS
BAR HIT
BAZ MISS
```
I'm curious if something like this has been explored already. Does this seem like it would work in general or am I missing something. Forgive me if I overlooked something in a prior discussion along these lines. 

Other affiliated issues/people: #9304 @luk-f-a @stuartarchibald

**Comment by stuartarchibald** [comment]
Many thanks for the detailed report @DannyWeitekamp. This will be discussed at the triage meeting tomorrow (2023-12-05). My initial thought is that this is similar/related to the issue of caching functions that are compiled from the exec of strings, a solution that works for one should work for the other. Creating a cache key based on the hash of the code object should work, the question is potentially how far-reaching (into the program state) such a cache key needs to be and what is a "sensible" behaviour. Whether this is generally preferable behaviour compared to file-at-a-time should probably be discussed more broadly with the Numba user base. Thanks again for looking into this!

**Comment by DannyWeitekamp** [comment]
So in my efforts to use this in my own projects a few additional considerations came up. 
1) Some special consideration needs to be given to `Dispatcher` and `_Intrinsic` dependencies. Just using cloudpickle 
to hash these doesn't generate a consistent hash, so some recursive accounting of dependencies is necessary. This would be easy if every Dispatcher had a _cache object, but that isn't always the case. Although it often works well enough just hashing one level out. 
2) Using this hash in `get_disambiguator` creates an nbi file for each unique recompile, which isn't ideal but honestly isn't much worse than using the line number which in the worst case creates several new indices in cases when you alter something high up in the file and all the line numbers change. It seems like the ideal disambiguator would be something like the n-th time a function with the same name was defined in a file... perhaps that 'n' could be determined by analyzing the file's byte code? Although perhaps that is not as easy as I think. 

I'm curious what the PIXIE solution would be for all this? Ideally the new AOT stuff would also have some kind of caching so individual pieces can recompile since it sounds like the plan is for there to be per-function compilation and reuse. I wouldn't mind it if I had to assign a fixed name and module to AOT stuff, and have jitted stuff be a little more cache-finnicky (especially if they compiled faster by reusing code). In this case function disambiguation, and recursive dependency tracking would be much more robust.

**Comment by dlee992** [comment]
looks like bazel also uses file-based cache? just FYI, [link](https://bazel.build/about/intro#:~:text=Bazel%20caches%20all%20previously%20done,highly%20parallel%20and%20incremental%20fashion.)
> Bazel caches all previously done work and tracks changes to both file content and build commands. This way, Bazel knows when something needs to be rebuilt, and rebuilds only that. To further speed up your builds, you can set up your project to build in a highly parallel and incremental fashion.
